src/components/stores/postgres-store.ts

```diff
@@ -0,0 +1,153 @@
+        opts.autocreateSchemaTable = opts.autocreateSchemaTable ?? true;
+        this.sql = opts.url ? postgres(opts.url, opts) : postgres(opts);
+        process.on("beforeExit", () => {
+            // Ensure we clean up on exit
+            this.destroy().catch(ex => {
+                log.warn('Failed to cleanly exit', ex);
+            });
+        })
+    }
+
+    /**
+     * Ensure the database schema is up to date. If you supplied
+     * `autocreateSchemaTable` to `opts` in the constructor, a fresh database
+     * will have a `schema` table created for it.
+     *
+     * @throws If a schema could not be applied cleanly.
+     */
+    public async ensureSchema(): Promise<void> {
+        log.info("Starting database engine");
+        let currentVersion = await this.getSchemaVersion();
+
+        if (currentVersion === -1) {
+            if (this.opts.autocreateSchemaTable) {
+                log.info(`Applying v0 schema (schema table)`);
+                await v0Schema(this.sql);
+                currentVersion = 0;
+            }
+        }
+        else {
+            // We aren't autocreating the schema table, so assume schema 0.
+            currentVersion = 0;
+        }
+
+        // Zero-indexed, so schema 1 would be in slot 0.
+        while (this.schemas[currentVersion]) {
+            log.info(`Updating schema to v${currentVersion + 1}`);
+            const runSchema = this.schemas[currentVersion];
+            try {
+                await runSchema(this.sql);
+                currentVersion++;
+                await this.updateSchemaVersion(currentVersion);
+            }
+            catch (ex) {
+                log.warn(`Failed to run schema v${currentVersion + 1}:`, ex);
+                throw Error("Failed to update database schema");
+            }
+        }
+        log.info(`Database schema is at version v${currentVersion}`);
+    }
+
+    /**
+     * Clean away any resources used by the database. This is automatically
+     * called before the process exits.
+     */
+    public async destroy(): Promise<void> {
+        log.info("Destroy called");
+        if (this.hasEnded) {
+            // No-op if end has already been called.
+            return;
+        }
+        this.hasEnded = true;
+        await this.sql.end();
+        log.info("PostgresSQL connection ended");
+    }
+
+    /**
+     * Update the current schema version.
+     * @param version
+     */
+    protected async updateSchemaVersion(version: number): Promise<void> {
+        log.debug(`updateSchemaVersion: ${version}`);
+        await this.sql`UPDATE schema SET version = ${version};`;
+    }
+
+    /**
+     * Get the current schema version.
+     * @returns The current schema version, or `-1` if no schema table is found.
+     */
+    protected async getSchemaVersion(): Promise<number> {
+        try {
+            const result = await this.sql<{version: number}[]>`SELECT version FROM SCHEMA;`;
+            return result?.[0]?.version;
```

Comment:
Does this need to check for `undefined`?  Or does it get caught by the `PostgresError`?
```suggestion
            return result?.[0]?.version ?? -1;
```

Reply:
Ah, this is probably safe tbh.